iceoryx_utils/include/iceoryx_utils/internal/lockfree_queue/index_queue.inl

```diff
@@ -0,0 +1,274 @@
+            }
+            // retry, readPosition was already updated via CAS in tryToGainOwnershipAt
+        }
+        else
+        {
+            // readPosition is ahead by one cycle, queue was empty at loadValueAt(...)
+            auto isEmpty = value.isOneCycleBehind(readPosition);
+
+            if (isEmpty)
+            {
+                // case (2)
+                return false;
+            }
+
+            // case (3) and (4) requires loading readPosition again
+            readPosition = loadNextReadPosition();
+        }
+
+        // readPosition is outdated, retry operation
+
+    } while (!ownershipGained); // we leave if we gain ownership of readPosition
+
+    index = value.getIndex();
+    return true;
+}
+
+template <uint64_t Capacity, typename ValueType>
+bool IndexQueue<Capacity, ValueType>::popIfFull(ValueType& index)
+{
+    // we do NOT need a CAS loop here since if we detect that the queue is not full
+    // someone else popped an element and we do not retry to check whether it was filled AGAIN
+    // concurrently (which will usually not be the case and then we would return false anyway)
+    // if it is filled again we can (and will) retry popIfFull from the call site
+
+    // the queue is full if and only if write position and read position are the same but read position is
+    // one cycle behind write position
+    // unfortunately it seems impossible in this design to check this condition without loading
+    // write posiion and read position (which causes more contention)
+
+    auto writePosition = loadNextWritePosition();
+    auto readPosition = loadNextReadPosition();
+    auto value = loadValueAt(readPosition);
+
+    auto isFull = writePosition.getIndex() == readPosition.getIndex() && readPosition.isOneCycleBehind(writePosition);
+
+    if (isFull)
+    {
+        if (tryToGainOwnershipAt(readPosition))
+        {
+            index = value.getIndex();
+            return true;
+        }
+    }
+
+    // otherwise someone else has dequeued an index and the queue was not full at the start of this popIfFull
+    return false;
+}
+
+template <uint64_t Capacity, typename ValueType>
+bool IndexQueue<Capacity, ValueType>::empty()
+{
+    auto oldReadIndex = m_readPosition.load(std::memory_order_relaxed);
+    auto value = m_cells[oldReadIndex.getIndex()].load(std::memory_order_relaxed);
+
+    // if m_readPosition is ahead by one cycle compared to the value stored at head,
+    // the queue was empty at the time of the loads above (but might not be anymore!)
+    return value.isOneCycleBehind(oldReadIndex);
+}
+
+template <uint64_t Capacity, typename ValueType>
+typename IndexQueue<Capacity, ValueType>::Index IndexQueue<Capacity, ValueType>::loadNextReadPosition() const
+{
+    return m_readPosition.load(std::memory_order_relaxed);
+}
+
+template <uint64_t Capacity, typename ValueType>
+typename IndexQueue<Capacity, ValueType>::Index IndexQueue<Capacity, ValueType>::loadNextWritePosition() const
```

Comment:
Please get rid of this unnecessary abstraction. Same argument as in loadNextReadPosition